src/DocumentFormat.OpenXml/Packaging/OpenXmlPartContainer.cs

```diff
@@ -1703,43 +1703,69 @@ internal OpenXmlPart AddSubPartFromOtherPackage(OpenXmlPart part, bool keepIdAnd
                     }
                 }
 
-                var updatedParts = new Dictionary<DataPart, DataPart>();
-                foreach (var item in dataPartsDictionary)
+                // Updates of existing dictionary keys during enumeration are only allowed after net 5.0.
+                // Before that we need a temporary dictionary to store the updated values for the keys.
+#if NET50_OR_LATER
+                UpdateDataPartsOptimized(dataPartsDictionary);
+#else
+                UpdateDataPartsUnoptimized(dataPartsDictionary);
+#endif
+
+                // then create data part reference relationship
+                foreach (var dataPartReferenceRelationship in part.DataPartReferenceRelationships)
                 {
-                    if (item.Value is null)
+                    if (dataPartsDictionary[dataPartReferenceRelationship.DataPart] is MediaDataPart newDataPart)
                     {
-                        var dataPart = item.Key;
-
-                        var newDataPart = new MediaDataPart(InternalOpenXmlPackage, dataPart.ContentType, dataPart.Uri);
+                        var newDataPartReference = DataPartReferenceRelationship.Create(child, newDataPart, dataPartReferenceRelationship.RelationshipType, dataPartReferenceRelationship.Id);
+                        child.AddDataPartReferenceRelationship(newDataPartReference);
+                    }
+                }
 
-                        // copy the stream
-                        using (var stream = dataPart.GetStream())
-                        {
-                            newDataPart.FeedData(stream);
-                        }
+                return child;
+            }
+        }
 
-                        InternalOpenXmlPackage.AddDataPartToList(newDataPart);
+        private void UpdateDataPartsUnoptimized(IDictionary<DataPart, DataPart?> dataPartsDictionary) {
```

Comment:
Can you have it be the same function name guarded by the `#if/def`?

Reply:
Do we have to take an `IDictionary<>` here? If not, enumerating is a bit faster on a `Dictionary<>` due to its custom enumerator

Reply:
Changed the `#if / def` and changed the functions to use a `Dictionary`